tools/document-pii/src/pii_finder.rs

```diff
@@ -0,0 +1,266 @@
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+
+use crate::EnumOrStruct;
+
+use anyhow::anyhow;
+use syn::visit::Visit;
+use syn::{Attribute, Field, ItemEnum, ItemStruct, Meta, Type};
+
+/// The name of a field along with its type. Used for the path to a pii-field.
+#[derive(Ord, PartialOrd, PartialEq, Eq, Hash, Clone, Debug, Default)]
+pub struct TypeAndField {
+    // Full path of a type. E.g. relay_common::protocol::Event, rather than just 'Event'.
+    pub qualified_type_name: String,
+    pub field_ident: String,
+}
+
+/// This is the visitor that actually generates the pii_types, it has a lot of associated data
+/// because it using the Visit trait from syn-crate means I cannot add data as arguments.
+/// The 'pii_types' field can be regarded as the output.
+pub struct PiiFinder<'a> {
+    pub module_path: String,
+    pub current_type: String,
+    pub all_types: &'a HashMap<String, EnumOrStruct>,
+    pub use_statements: &'a BTreeMap<String, BTreeSet<String>>,
+    pub pii_values: &'a Vec<String>,
+    pub current_path: &'a mut Vec<TypeAndField>,
+    pub pii_types: &'a mut BTreeSet<Vec<TypeAndField>>, // output
+}
+
+impl<'a> PiiFinder<'a> {
+    /// Takes a Field and visit the types that it consist of if we have
+    /// the full path to it in self.use_statements.
+    fn visit_field_types(&mut self, node: &Field) {
+        let local_paths = self.use_statements.get(&self.module_path).unwrap().clone();
+
+        let mut field_types = vec![];
+        get_field_types(&node.ty, &mut field_types);
+
+        for field_type in &field_types {
+            for use_path in &local_paths {
+                if use_path.split("::").last().unwrap() == field_type.trim() {
+                    let enum_or_struct = { self.all_types.get(use_path).cloned() };
+
+                    if let Some(enum_or_struct) = enum_or_struct {
+                        match enum_or_struct {
+                            EnumOrStruct::Enum(itemenum) => {
+                                let current_type = self.current_type.clone();
+                                let module_path = self.module_path.clone();
```

Comment:
Here we have 6 indentation levels too. In general, don't be afraid to split your code into functions, even if it's only called once at this point. It makes the code easier to read, understand, extend, and test.

Reply:
revised it now